File: n8n/scripts/alumas_whatsapp_campaign.py
#!/usr/bin/env python3
import argparse
import json
import sys
from datetime import datetime
from pathlib import Path
from zoneinfo import ZoneInfo


DAY_NAMES = ["lunes", "martes", "miércoles", "jueves", "viernes", "sábado", "domingo"]


def week_bucket(day_of_month: int) -> int:
    if 1 <= day_of_month <= 8:
        return 1
    if 9 <= day_of_month <= 16:
        return 2
    if 17 <= day_of_month <= 24:
        return 3
    return 4


def load_contacts(path: Path) -> dict:
    with path.open("r", encoding="utf-8") as handle:
        return json.load(handle)


def normalize_day(value: str) -> str:
    lowered = value.strip().lower()
    aliases = {
        "miercoles": "miércoles",
        "sabado": "sábado",
    }
    return aliases.get(lowered, lowered)


def build_campaign(data: dict, timezone_name: str, forced_day: str | None = None) -> dict:
    now = datetime.now(ZoneInfo(timezone_name))
    today = DAY_NAMES[now.weekday()]
    selected_day = normalize_day(forced_day) if forced_day else today

    mensajes_por_dia = data.get("mensajes_por_dia", {})
    promociones_por_semana = data.get("promociones_por_semana", {})
    selected_week = week_bucket(now.day)
    promotion = promociones_por_semana.get(str(selected_week), "") or promociones_por_semana.get(selected_week, "")
    contacts = mensajes_por_dia.get(selected_day, [])

    return {
        "timezone": timezone_name,
        "generated_at": now.isoformat(),
        "system_day": today,
        "selected_day": selected_day,
        "day_matches_system": selected_day == today,
        "selected_week_bucket": selected_week,
        "promotion": promotion,
        "contact_count": len(contacts),
        "contacts": [
            {
                "phone": item[0],
                "message": item[1],
            }
            for item in contacts
            if isinstance(item, list) and len(item) >= 2
        ],
    }


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Resuelve la campaña diaria de WhatsApp de ALUMAS para n8n."
    )
    parser.add_argument("--contacts-path", required=True, help="Ruta absoluta a contactos.json")
    parser.add_argument(
        "--timezone",
        default="America/Bogota",
        help="Zona horaria IANA para resolver el día activo",
    )
    parser.add_argument(
        "--force-day",
        default="",
        help="Sobrescribe el día calculado. Solo para validaciones controladas.",
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="No intenta enviar. Solo devuelve el plan de campaña.",
    )
    parser.add_argument(
        "--output-path",
        default="",
        help="Opcional. Guarda el resultado JSON en disco además de imprimirlo.",
    )
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    contacts_path = Path(args.contacts_path).expanduser().resolve()
    if not contacts_path.exists():
        print(
            json.dumps(
                {
                    "ok": False,
                    "error": f"No existe el archivo de contactos: {contacts_path}",
                },
                ensure_ascii=False,
            )
        )
        return 1

    try:
        data = load_contacts(contacts_path)
        campaign = build_campaign(
            data=data,
            timezone_name=args.timezone,
            forced_day=args.force_day or None,
        )
    except Exception as exc:
        print(
            json.dumps(
                {
                    "ok": False,
                    "error": str(exc),
                },
                ensure_ascii=False,
            )
        )
        return 1

    payload = {
        "ok": True,
        "mode": "dry-run" if args.dry_run else "plan-only",
        "transport_ready": False,
        "transport_message": (
            "La capa de envio Linux aun no esta implementada. "
            "Este script resuelve el dia correcto, la promocion vigente y el lote del dia."
        ),
        **campaign,
    }

    rendered = json.dumps(payload, ensure_ascii=False, indent=2)
    if args.output_path:
        output_path = Path(args.output_path).expanduser().resolve()
        output_path.parent.mkdir(parents=True, exist_ok=True)
        output_path.write_text(rendered + "\n", encoding="utf-8")

    print(rendered)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: n8n/scripts/alumas_whatsapp_campaign.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');

const DAY_NAMES = ['lunes', 'martes', 'miércoles', 'jueves', 'viernes', 'sábado', 'domingo'];

function parseArgs(argv) {
  const args = {};
  for (let i = 2; i < argv.length; i += 1) {
    const token = argv[i];
    if (!token.startsWith('--')) continue;
    const key = token.slice(2);
    const next = argv[i + 1];
    if (!next || next.startsWith('--')) {
      args[key] = true;
      continue;
    }
    args[key] = next;
    i += 1;
  }
  return args;
}

function normalizeDay(value) {
  const lowered = String(value || '').trim().toLowerCase();
  const aliases = {
    miercoles: 'miércoles',
    sabado: 'sábado',
  };
  return aliases[lowered] || lowered;
}

function weekBucket(dayOfMonth) {
  if (dayOfMonth >= 1 && dayOfMonth <= 8) return 1;
  if (dayOfMonth >= 9 && dayOfMonth <= 16) return 2;
  if (dayOfMonth >= 17 && dayOfMonth <= 24) return 3;
  return 4;
}

function nowInTimezone(timezone) {
  const parts = new Intl.DateTimeFormat('en-CA', {
    timeZone: timezone,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: false,
    weekday: 'long',
  }).formatToParts(new Date());
  const byType = Object.fromEntries(parts.map((part) => [part.type, part.value]));
  const englishWeekday = (byType.weekday || '').toLowerCase();
  const weekdayMap = {
    monday: 'lunes',
    tuesday: 'martes',
    wednesday: 'miércoles',
    thursday: 'jueves',
    friday: 'viernes',
    saturday: 'sábado',
    sunday: 'domingo',
  };

  return {
    isoLocal: `${byType.year}-${byType.month}-${byType.day}T${byType.hour}:${byType.minute}:${byType.second}`,
    dayOfMonth: Number(byType.day),
    weekday: weekdayMap[englishWeekday] || englishWeekday,
  };
}

function normalizePhone(rawPhone) {
  let digits = String(rawPhone || '').replace(/\D/g, '');
  digits = digits.replace(/^0+/, '');
  if (digits.length === 10 && digits.startsWith('3')) {
    digits = `57${digits}`;
  }
  if (digits.length === 12 && digits.startsWith('57') && digits[2] === '3') {
    return digits;
  }
  return digits || null;
}

function appendLog(logPath, payload) {
  if (!logPath) return;
  fs.mkdirSync(path.dirname(logPath), { recursive: true });
  fs.appendFileSync(logPath, `${JSON.stringify(payload, null, 0)}\n`, 'utf8');
}

function loadContacts(contactsPath) {
  return JSON.parse(fs.readFileSync(contactsPath, 'utf8'));
}

function buildCampaign(data, timezone, forcedDay) {
  const now = nowInTimezone(timezone);
  const selectedDay = forcedDay ? normalizeDay(forcedDay) : now.weekday;
  const mensajesPorDia = data.mensajes_por_dia || {};
  const promocionesPorSemana = data.promociones_por_semana || {};
  const selectedWeekBucket = weekBucket(now.dayOfMonth);
  const promotion = promocionesPorSemana[String(selectedWeekBucket)] || promocionesPorSemana[selectedWeekBucket] || '';
  const rawContacts = Array.isArray(mensajesPorDia[selectedDay]) ? mensajesPorDia[selectedDay] : [];

  const contacts = rawContacts
    .filter((item) => Array.isArray(item) && item.length >= 2)
    .map((item) => ({
      phone: normalizePhone(item[0]),
      originalPhone: item[0],
      message: String(item[1] || '').trim(),
      raw: item,
    }))
    .filter((item) => item.phone && item.message);

  return {
    timezone,
    generated_at: now.isoLocal,
    system_day: now.weekday,
    selected_day: selectedDay,
    day_matches_system: selectedDay === now.weekday,
    selected_week_bucket: selectedWeekBucket,
    promotion,
    contact_count: contacts.length,
    contacts,
  };
}

function composeMessage(baseMessage, promotion) {
  const cleanBase = String(baseMessage || '').trim();
  const cleanPromotion = String(promotion || '').trim();
  if (!cleanPromotion) return cleanBase;
  if (!cleanBase) return cleanPromotion;
  return `${cleanBase}\n\n${cleanPromotion}`;
}

async function sendMessage({ baseUrl, instanceName, instanceToken, number, text }) {
  const response = await fetch(`${baseUrl.replace(/\/$/, '')}/message/sendText/${instanceName}`, {
    method: 'POST',
    headers: {
      apikey: instanceToken,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      number,
      text,
    }),
  });

  const raw = await response.text();
  let parsed;
  try {
    parsed = raw ? JSON.parse(raw) : {};
  } catch {
    parsed = { raw };
  }

  return {
    ok: response.ok,
    status: response.status,
    body: parsed,
  };
}

async function main() {
  const args = parseArgs(process.argv);
  const contactsPath = path.resolve(String(args['contacts-path'] || ''));
  const timezone = String(args.timezone || 'America/Bogota');
  const forcedDay = args['force-day'] ? String(args['force-day']) : '';
  const dryRun = Boolean(args['dry-run']);
  const logPath = args['log-path'] ? path.resolve(String(args['log-path'])) : '';
  const outputPath = args['output-path'] ? path.resolve(String(args['output-path'])) : '';
  const delayMs = Number(args['delay-ms'] || 1200);
  const maxMessages = args.limit ? Number(args.limit) : 0;
  const baseUrl = String(args['evolution-base-url'] || '');
  const instanceName = String(args['instance-name'] || '');
  const instanceToken = String(args['instance-token'] || '');

  if (!contactsPath || !fs.existsSync(contactsPath)) {
    console.log(JSON.stringify({ ok: false, error: `No existe el archivo de contactos: ${contactsPath}` }, null, 2));
    process.exit(1);
  }

  let data;
  let campaign;
  try {
    data = loadContacts(contactsPath);
    campaign = buildCampaign(data, timezone, forcedDay);
  } catch (error) {
    console.log(JSON.stringify({ ok: false, error: String(error.message || error) }, null, 2));
    process.exit(1);
  }

  const basePayload = {
    ok: true,
    mode: dryRun ? 'dry-run' : 'send',
    transport_ready: Boolean(baseUrl && instanceName && instanceToken),
    ...campaign,
  };

  if (!campaign.day_matches_system && !forcedDay) {
    const payload = {
      ...basePayload,
      ok: false,
      error: `El día calculado (${campaign.selected_day}) no coincide con el día del sistema (${campaign.system_day}).`,
    };
    console.log(JSON.stringify(payload, null, 2));
    process.exit(1);
  }

  const selectedContacts = maxMessages > 0 ? campaign.contacts.slice(0, maxMessages) : campaign.contacts;

  if (dryRun || !selectedContacts.length) {
    const payload = {
      ...basePayload,
      contacts: selectedContacts,
      send_results: [],
      sent_count: 0,
      failed_count: 0,
      skipped_count: campaign.contact_count - selectedContacts.length,
    };
    if (outputPath) {
      fs.mkdirSync(path.dirname(outputPath), { recursive: true });
      fs.writeFileSync(outputPath, `${JSON.stringify(payload, null, 2)}\n`, 'utf8');
    }
    console.log(JSON.stringify(payload, null, 2));
    return;
  }

  if (!baseUrl || !instanceName || !instanceToken) {
    console.log(JSON.stringify({ ...basePayload, ok: false, error: 'Faltan parámetros de transporte para Evolution API.' }, null, 2));
    process.exit(1);
  }

  const sendResults = [];
  let sentCount = 0;
  let failedCount = 0;

  for (let index = 0; index < selectedContacts.length; index += 1) {
    const contact = selectedContacts[index];
    const text = composeMessage(contact.message, campaign.promotion);
    const attempt = await sendMessage({
      baseUrl,
      instanceName,
      instanceToken,
      number: contact.phone,
      text,
    });

    const result = {
      index,
      phone: contact.phone,
      status: attempt.status,
      ok: attempt.ok,
      remoteJid: attempt.body?.key?.remoteJid || null,
      messageId: attempt.body?.key?.id || null,
    };
    sendResults.push(result);

    appendLog(logPath, {
      timestamp: new Date().toISOString(),
      selected_day: campaign.selected_day,
      phone: contact.phone,
      ok: attempt.ok,
      status: attempt.status,
      messageId: result.messageId,
    });

    if (attempt.ok) {
      sentCount += 1;
    } else {
      failedCount += 1;
    }

    if (index < selectedContacts.length - 1 && delayMs > 0) {
      await new Promise((resolve) => setTimeout(resolve, delayMs));
    }
  }

  const payload = {
    ...basePayload,
    contacts: selectedContacts,
    send_results: sendResults,
    sent_count: sentCount,
    failed_count: failedCount,
    skipped_count: campaign.contact_count - selectedContacts.length,
  };

  if (outputPath) {
    fs.mkdirSync(path.dirname(outputPath), { recursive: true });
    fs.writeFileSync(outputPath, `${JSON.stringify(payload, null, 2)}\n`, 'utf8');
  }

  console.log(JSON.stringify(payload, null, 2));
}

main().catch((error) => {
  console.log(JSON.stringify({ ok: false, error: String(error.message || error) }, null, 2));
  process.exit(1);
});
